specs/CustomDownload.md

```diff
@@ -0,0 +1,374 @@
+# Background
+<!-- TEMPLATE
+    Use this section to provide background context for the new API(s)
+    in this spec.
+
+    This section and the appendix are the only sections that likely
+    do not get copied into any official documentation, they're just an aid
+    to reading this spec.
+
+    If you're modifying an existing API, included a link here to the
+    existing page(s) or spec documentation.
+
+    For example, this section is a place to explain why you're adding this
+    API rather than modifying an existing API.
+
+    For example, this is a place to provide a brief explanation of some dependent
+    area, just explanation enough to understand this new API, rather than telling
+    the reader "go read 100 pages of background information posted at ...".
+-->
+There have been multiple requests by WebView Developers to be able to customize/control the download experience within WebView2.
+
+Some of the requests we've heard, and want to solve with this API are:
+- Developers want to be able to disable downloads for the entire environment.
+- Developers want to be able to disable downloads on a per-case basis. For example, based on the content type, or user.
+- Developers want to have access to downloaded file’s metadata.
+- Developers want to modify the download path.
+- Developers don’t want to see edge download UI if download is cancelled.
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Description
+
+Our proposed solution is to give developers an API to control the download experience.
+We will do this by exposing a DownloadStarting event.
+
+The DownloadStarting event will be raised when a download has begun.
+The host can then choose to cancel a download, change the save path, and hide the default download dialog.
```

Comment:
If we're hiding the default UI download dialog of Edge. How we're informed about the download progress then?! Let's assume we have a 4GB file to download and we would like to hide the Edge download UI and display our own instead. How can we tell the user how much bytes have been downloaded already? From what I see we cannot, the download will be executed in background and nobody but Edge knows the progress and when its done.

Reply:
I feel the same. For a complete experience we would also need a DownloadProgress event with the amount of bytes received up to that point.

Second, but I'm not that familiar with `Windows.Foundation.Deferral`, it reads that it only signals on download completed. How can be distinguished if the download completed succesfully, or failed somewhere?

If it failed we would also need a reason. Connection lost, out of diskspace, ...

Reply:
Thanks @symbai and @frankdekker, both of you are correct, just blocking the download dock would leave a question mark about download progress. We intend to release a download progress API, although we are still early in the api drafting process, which is why it is not included in this PR. That said, we have heard customers request a way to simply disable the UI, which this API would solve.

Frank, to your second question: would a download completed event be useful? This is a valid request, and I can add it to our backlog. And to your 3rd question, the oppurtunity to cancel the download programmatically would happen when you intercept the downloadStarting event, do you expect to cancel the download otherwise?

Reply:
@jasonstephen15 
I'm reasoning from the perspective that I can build a custom download GUI. To present the user a fully functional gui I need to know the following events:

- DownloadStartingEvent - where I can prompt the user if he wants to open the download it or save it at a specified location.
- DownloadProgressEvent - where I can show a download progress indicator somewhere in my application.
   -  this should contain info about the bytesReceived and pref about the transfer speed.
- DownloadCompletedEvent - where i can tell the user that the download finished
   - this should contain info about if it completed or failed.

Additionally a user could accept the download, but if it takes too long, still wants to choose to abort the download. The API should provide a mechanisme to abort the download. I assume cancelling the DownloadStartingEvent only allows the ability to cancel it at the start. For complete functionality we should also be able to cancel it _after_ the DownloadStartingEvent.

As an example the API that I'm used to in a Java environment:

DownloadStartingEvent contains a DownloadItem object and contains the following:
- event DownloadProgress
- event DownloadCompleted
- event DownloadFailed
- void Cancel()

As developer this provides me with the possibility to control every aspect of a single item download.